New Import UI feature - WIP

seando-adsk opened the pull request:
#### MAYA-101325 - Implement Qt Model base class that supports/exposes a USD stage
#### MAYA-101326 - Implement filtering proxy on top of USD stage model
#### MAYA-101566 - Hook up the new Import pop-up dialog options to import command
#### MAYA-101565 - Add VariantSets to import dialog and allow selection of a Variant

* Added new build flag --qt-location with corresponding define QT_LOCATION.
* Building of Qt UI is optional based on 'CMAKE_WANT_QT_BUILD' (default ON) and QT_LOCATION being passed in.
* Added optional SUBDIR arg when promoting files.
* Created a USD ImportUI dialog from internal design.
* Added usd/ui folder with new ImportUI dialog.
* From the mayaUsd plugin register a new command to open the ImportUI dialog.
* Added the mayaUsd translator import options based on design specs.

* New import data, which supports:
** Root prim path.
** Stage population mask for importing partial usd file.
** Stage load set.
** Variant sets associated with root prim and invididual prims.
** Added clearData flag to usdImportDialog command.
** Merged variants into single column.
** Added ItemDelegate for customized widget creation for Variants. Use persistent editor with label/combobox for variant selection.
** Moved tree item creation logic from factory to tree item.
** Tree checking & stage population mask.
** New tree header with checkbox for load column - not hooked up yet.
** Commented out all the usd import options (for now) as none of them are implemented yet.

seando-adsk commented on build.py:
```diff
@@ -371,6 +375,9 @@ def RunTests(context,extraArgs):
 
+parser.add_argument("--qt-location", type=str,
+                    help="Directory where Qt is installed.")
+
```
To build the new UI (which uses) Qt you need to specify where to find Qt (should be the same one that Maya is using).

seando-adsk commented on cmake/modules/FindMaya.cmake:
```diff
@@ -169,19 +168,6 @@ find_path(MAYA_INCLUDE_DIR
         "Maya's headers path"
 )
 
-find_path(MAYA_LIBRARY_DIR
```
This is not needed. It is duplicated from above in each platform specific section.

seando-adsk commented on lib/fileio/importData.cpp:
```diff
@@ -0,0 +1,146 @@
+//
```
New file containing all the import data, such as root prim path, stage pop mask, variants, etc.

seando-adsk commented on lib/fileio/jobs/readJob.cpp:
```diff
@@ -60,14 +60,12 @@ PXR_NAMESPACE_OPEN_SCOPE
         const std::string &iFileName,
-        const std::string &iPrimPath,
-        const std::map<std::string, std::string>& iVariants,
+        const MayaUsd::ImportData &iImportData,
```
You'll see a few changes like this. I've replaced the primPath and variants arg, with my new import data (which contains both those and more).

seando-adsk commented on lib/fileio/jobs/readJob.cpp:
```diff
@@ -103,7 +102,22 @@ UsdMaya_ReadJob::Read(std::vector<MDagPath>* addedDagPaths)
     UsdStageCacheContext stageCacheContext(UsdMayaStageCache::Get());
-    UsdStageRefPtr stage = UsdStage::Open(rootLayer, sessionLayer);
+    UsdStageRefPtr stage;
+    if (mImportData.empty())
```
Change to how we open the stage depending on what the import data contains.

seando-adsk replied:
Yes correct, now that I've moved the filename into the import data it won't be empty. I've changed the stage open as you've suggested.

seando-adsk commented on lib/fileio/jobs/readJob.cpp:
```diff
@@ -176,12 +191,22 @@ UsdMaya_ReadJob::Read(std::vector<MDagPath>* addedDagPaths)
+                .GetVariantSet(variant.first)
+                .SetVariantSelection(variant.second);
+        }
+    }
```
New loop to set all the individual prim variants.

seando-adsk commented on lib/usd/CMakeLists.txt:
```diff
@@ -1 +1,10 @@
+    else()
+        message(STATUS "QT_LOCATION not set. Building UI features will be disabled.")
+    endif()
+endif()
```
When the new cmake option is on (enabled by default) and we got a QT_LOCATION passed to build, we'll add the new "ui" subdir.

seando-adsk commented on lib/usd/ui/api.h:
```diff
@@ -0,0 +1,47 @@
+//
```
New api file - similar to the one from lib folder.

seando-adsk commented on lib/usd/ui/CMakeLists.txt:
```diff
@@ -0,0 +1,131 @@
+if (BUILD_TESTS)
+	# A simple executable used to test the import dialog.
+	add_subdirectory(demo)
+endif()
```
This builds a standalone executable that opens the import ui dialog. Useful during development.

seando-adsk commented on lib/usd/ui/cmds/USDImportDialogCmd.cpp:
```diff
@@ -0,0 +1,122 @@
+	MStatus st;
+	MArgParser argData(syntax(), args, &st);
+
+	if(argData.isFlagSet(kClearDataFlag))
```
The clear data flag is called by the import dialog when you click on a new file.

seando-adsk commented on lib/usd/ui/cmds/USDImportDialogCmd.cpp:
```diff
@@ -0,0 +1,122 @@
+	syntax.enableQuery(false);
+	syntax.enableEdit(false);
+	syntax.addFlag(kClearDataFlag, kClearDataFlagLong);
+	syntax.setObjectType(MSyntax::kStringObjects, 1, 1);
```
The command expects a file path to the USD file.

seando-adsk commented on lib/usd/ui/demo/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+
+add_executable(${PROJECT_NAME}
+	testMayaUsdUI.cpp
+)
```
Simple text application to open the dialog. So I can test as standalone without Maya.

seando-adsk commented on lib/usd/ui/views/ItemDelegate.cpp:
```diff
@@ -0,0 +1,260 @@
+//
```
Item delegate used to customize the drawing of the variant columns.

seando-adsk commented on lib/usd/ui/views/TreeHeader.cpp:
```diff
@@ -0,0 +1,76 @@
+//
```
To have a checkbox in the header.

seando-adsk commented on lib/usd/ui/views/TreeItem.h:
```diff
@@ -0,0 +1,104 @@
+ * displayed to the User when interacting with Tree content.
+ */
+class MAYAUSD_UI_PUBLIC TreeItem : public QStandardItem
+{
```
Note: there is no Q_OBJECT here as QStandardItem does not derive from QObject. So there cannot be signals/slots in this class.

seando-adsk commented on lib/usd/ui/views/TreeModel.cpp:
```diff
@@ -0,0 +1,239 @@
+	emit dataChanged(rMinIndex, rMaxIndex);
+}
+
+void TreeModel::fillStagePopulationMask(UsdStagePopulationMask& popMask, const QModelIndex& parent)
```
Helper to fill the stage pop mask depending on what is checked.

seando-adsk commented on lib/usd/ui/views/TreeModel.cpp:
```diff
@@ -0,0 +1,239 @@
+	return ParentClass::setData(index, value, role);
+}
+
+void TreeModel::setParentsCheckState(const QModelIndex &child, Qt::CheckState state)
```
This method (and next one) are used to check/uncheck items parent/child items when the user interactively checks/unchecks an item in the tree.

seando-adsk commented on lib/usd/ui/views/TreeModel.cpp:
```diff
@@ -0,0 +1,239 @@
+	}
+}
+
+void TreeModel::openPersistentEditors(QTreeView* tv, const QModelIndex& parent)
```
Helper called after tree is created to create all the variant set widgets on each row (that has variant sets).

In Qt terms this is called a persistent editor. Meaning it's always there, rather than only being created when you click in a cell.

seando-adsk commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,225 @@
+	// WIP - the search buttons are not implemented yet, so disable them.
+	fUI->findPrev->setHidden(true);
+	fUI->findNext->setHidden(true);
+	fUI->findAll->setHidden(true);
```
WIP - I hid these buttons for now as they aren't implemented yet.

mattyjams replied:
I think it would be preferable for work-in-progress functionality to **not** be included in a PR. When the functionality is ready, a PR dedicated to adding the new buttons and the code that drives them becomes easier to review, as does this PR.

seando-adsk commented on lib/usd/ui/views/USDImportDialog.ui:
```diff
@@ -0,0 +1,278 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Created as output from Qt designer.

seando-adsk commented on plugin/adsk/plugin/CMakeLists.txt:
```diff
@@ -43,6 +47,13 @@ target_link_libraries(${PLUGIN_PACKAGE}
+if(CMAKE_WANT_QT_BUILD AND DEFINED QT_LOCATION)
+    target_link_libraries(${PLUGIN_PACKAGE}
+        PRIVATE
+            mayaUsdUI
```
This plugin needs the new UI lib to register the command.

seando-adsk commented on plugin/adsk/scripts/mayaUsdTranslatorImport.mel:
```diff
@@ -42,16 +43,93 @@ global proc int mayaUsdTranslatorImport (string $parent,
+        rowLayout -numberOfColumns 3 -cw 1 $cw1 -cat 1 "right" 0;
+            text -label "Scope and Variants: " -al "right";
+            textField -ed false -nbg true -text "Please select a file to enable this option." -w $cw2 mayaUsdTranslator_SelectFileField;
+            button -label "View Hierarchy" -c "string $usdFile = currentFileDialogSelection(); usdImportDialog $usdFile;" mayaUsdTranslator_ViewHierBtn;
```
The button for opening dialog.

seando-adsk commented on plugin/adsk/scripts/mayaUsdTranslatorImport.mel:
```diff
@@ -42,16 +43,93 @@ global proc int mayaUsdTranslatorImport (string $parent,
+//
+//        checkBoxGrp -label "Animation Data: " -cw 1 $cw1 mayaUsdTranslator_AnimDataCheckBox;
+//
+//        intFieldGrp -label "Custom Frame Range Start/End: " -nf 2 -cw 1 $cw2 mayaUsdTranslator_CustomFrameRange;
```
The rest of the options area here (which I showed at demo). But are all commented for now as they aren't implemented yet.

seando-adsk commented on plugin/adsk/scripts/mayaUsdTranslatorImport.mel:
```diff
@@ -42,16 +43,93 @@ global proc int mayaUsdTranslatorImport (string $parent,
 
         eval($resultCallback+" \""+$currentOptions+"\"");
 
+    } else if ($action == "fileselected") {
```
This new section is the callback when user selects something in import dialog. I use it to hide/show the button.

seando-adsk commented:
@murphyeoin Would anyone at Animal Logic be interested in reviewing this?

@mattyjams I modified the Pixar plugin, so I included you.

@elrond79 A lot of new Qt code, would you like to review?

mattyjams commented on lib/usd/ui/views/ItemDelegate.h:
```diff
@@ -0,0 +1,106 @@
+
+	// The variant selection role is used to hold the current variant selection
+	// (from the qt combobox).
+	static constexpr int kVariantSelectionRole = kVariantNameRole + 1;
```
Is an enum more appropriate here?

mattyjams commented on lib/usd/ui/views/TreeHeader.h:
```diff
@@ -0,0 +1,47 @@
+//
+#pragma once
+
+#include "../api.h"
```
Relative header include.

mattyjams commented on lib/usd/ui/views/TreeItem.cpp:
```diff
@@ -0,0 +1,90 @@
+
+void TreeItem::setCheckState(Qt::CheckState st)
+{
+	assert(fType == Type::kLoad);
```
More asserts. I think we probably want to avoid those in general. `TF_VERIFY()` might be a better option here too.

mattyjams commented on lib/usd/ui/views/TreeItem.h:
```diff
@@ -0,0 +1,104 @@
+//
+#pragma once
+
+#include "../api.h"
```
Relative header include.

mattyjams commented on lib/usd/ui/views/TreeModel.h:
```diff
@@ -0,0 +1,78 @@
+//
+#pragma once
+
+#include "../api.h"
```
Relative header include.

mattyjams commented on lib/usd/ui/views/TreeModel.h:
```diff
@@ -0,0 +1,78 @@
+		kTreeColumn_Load,				// Should we load this prim?
+		kTreeColumn_Name,				// Name of the item as it appears in the TreeView.
+		kTreeColumn_Type,				// Type of the primitive.
+		kTreeColumn_Variants,			// Variant Set(s) and Variant Selection of the primitive.
```
Funky whitespace.

mattyjams commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,245 @@
+
+#include "USDImportDialog.h"
+#include "factories/TreeModelFactory.h"
+//#include "TreeHeader.h"
```
Remove?

seando-adsk replied:
I prefer not to. This is work in progress that goes along with commented code at line 73.

mattyjams commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,245 @@
+
+MAYAUSD_NS_DEF {
+
+IUSDImportView::~IUSDImportView() { }
```
Why put this here? Seems like it should just be declared `= default` in the header.

mattyjams commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,245 @@
+	return fFilename;
+}
+
+std::string USDImportDialog::rootPrimPath() const
```
Return a const reference?

mattyjams commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,245 @@
+	fSearchTimer->setSingleShot(true);
+	connect(fSearchTimer.get(), &QTimer::timeout, fSearchTimer.get(), [this]() {
+		fUI->treeView->setEnabled(false);
+//		fOverlay->StartSpinning();
```
Remove?

mattyjams commented on lib/usd/ui/views/USDImportDialog.cpp:
```diff
@@ -0,0 +1,245 @@
+// 		else
+// 		{
+// 			fOverlay->ShowInformationMessage(tr("Your search did not match any Prim."));
+// 		}
```
Remove?

mattyjams commented on lib/usd/ui/views/USDImportDialog.h:
```diff
@@ -0,0 +1,112 @@
+//
+#pragma once
+
+#include "../api.h"
```
Relative header include.

mattyjams commented on lib/usd/ui/views/USDImportDialog.h:
```diff
@@ -0,0 +1,112 @@
+#include <pxr/usd/usd/stage.h>
+
+#include "IUSDImportView.h"
+// #include "SpinnerOverlayWidget.h"
```
Remove?

mattyjams commented on lib/usd/ui/views/USDImportDialog.h:
```diff
@@ -0,0 +1,112 @@
+	mutable std::string fRootPrimPath;
+
+// 	// TreeView overlay on which to display an animated Spinner or message to the User:
+// 	std::unique_ptr<SpinnerOverlayWidget> fOverlay;
```
Remove?

mattyjams commented on lib/usd/ui/views/USDSearchThread.h:
```diff
@@ -0,0 +1,69 @@
+//
+#pragma once
+
+#include "../api.h"
```
Relative header include.

mattyjams commented on plugin/adsk/plugin/importTranslator.cpp:
```diff
@@ -113,7 +113,10 @@ UsdMayaImportTranslator::reader(
-    UsdMaya_ReadJob mUsdReadJob(fileName, primPath, variants, jobArgs);
+    UsdMaya_ReadJob mUsdReadJob(
+        fileName,
+        MayaUsd::ImportData::instance(),
```
Seems like we should be constructing an ImportData from the arguments rather than pulling from a singleton.

mattyjams commented on plugin/adsk/scripts/mayaUsdTranslatorImport.mel:
```diff
@@ -41,20 +42,165 @@ global proc int mayaUsdTranslatorImport (string $parent,
+//            menuItem "All";
+//            menuItem "None";
+//            menuItem "Custom";
+//        optionMenuGrp -e -sl 2 mayaUsdTranslator_IncludeCustomAttribOptionMenu;
```
Remove all of this?

mattyjams commented on plugin/pxr/maya/lib/usdMaya/importCommand.cpp:
```diff
@@ -201,7 +201,11 @@ UsdMayaImportCommand::doIt(const MArgList & args)
 
-    mUsdReadJob = new UsdMaya_ReadJobWithSceneAssembly(mFileName, mPrimPath, mVariants, jobArgs);
+    MayaUsd::ImportData importData;
+    importData.setRootVariantSelections(std::move(mVariants));
```
Add mFileName to the importData too?

mattyjams commented on plugin/pxr/maya/lib/usdMaya/importTranslator.cpp:
```diff
@@ -111,8 +109,9 @@ UsdMayaImportTranslator::reader(
             userArgs,
             /* importWithProxyShapes = */ false,
             timeInterval);
+    MayaUsd::ImportData importData;
```
Set the filename on the importData?

mattyjams commented on plugin/pxr/maya/lib/usdMaya/referenceAssembly.cpp:
```diff
@@ -1530,9 +1530,13 @@ bool UsdMayaRepresentationHierBase::activate()
                 GfInterval::GetFullInterval());
+
+    MayaUsd::ImportData importData;
+    importData.setRootVariantSelections(std::move(variantSetSelections));
```
Add usdFilePath to importData?

seando-adsk commented on lib/usd/ui/CMakeLists.txt:
```diff
@@ -0,0 +1,131 @@
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+set(CMAKE_PREFIX_PATH "${QT_LOCATION}")
```
Your right sorry I do not need "CMAKE_INCLUDE_CURRENT_DIR". That is only needed for older cmake versions.
But without CMAKE_PREFIX_PATH the find_package(Qt5) will fail because it does not know where Qt is located. I followed what is suggested here: https://doc.qt.io/qt-5/cmake-get-started.html#build-a-gui-executable
Although I set it directly here rather than in an env var.

seando-adsk commented on lib/fileio/importData.h:
```diff
@@ -0,0 +1,129 @@
+	VariantSelections			fRootVariants;
+	PrimVariantSelections		fPrimVariants;
+	std::string					fRootPrimPath;
+	std::string					fFilename;
```
I'm a "tab" guy so all my new files use tabs. If you set your tabsize=4 this all lines up nicely.

seando-adsk commented on lib/usd/ui/cmds/USDImportDialogCmd.cpp:
```diff
@@ -0,0 +1,122 @@
+/*static*/
+MStatus USDImportDialogCmd::initialize(MFnPlugin& plugin)
+{
+	return plugin.registerCommand(
```
It's all tabs (1 on line 56 and 2 on lines 57-59).

seando-adsk commented on lib/usd/ui/demo/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+)
+
+target_include_directories(${PROJECT_NAME}
+	PUBLIC
```
1-tab

seando-adsk commented on lib/usd/ui/factories/TreeModelFactory.h:
```diff
@@ -0,0 +1,145 @@
+	 * \param parent A reference to the parent of the TreeModel.
+	 * \return A TreeModel created from the given USD Stage.
+	 */
+	static std::unique_ptr<TreeModel> createFromStage(const UsdStageRefPtr& stage, QObject* parent = nullptr, int* nbItems = nullptr);
```
I see how an argument could be made for size_t. I made it an int because it is eventually used in USDImportDialog.cpp as an arg to a QString (to replace the %1). QString arg method uses int. So I just used int. If I switched it to size_t from the method then I'll have to cast to an int for the QString. What do you think?

I did forget to add to the doc string. I'll do that.

HamedSabri-adsk commented on cmake/utils.cmake:
```diff
@@ -156,10 +156,36 @@ function(mayaUsd_promoteMayaUsdHeader)
+    foreach(header ${headerFiles})
         set(srcFile ${CMAKE_CURRENT_SOURCE_DIR}/${header})
-        set(dstFile ${CMAKE_BINARY_DIR}/include/mayaUsd/${header})
+        set(dstFile ${DEST_DIR}/${header})
```
@mattyjams great point. I am going to have a separate PR addressing all that after this PR  is merged into dev.

seando-adsk commented on lib/usd/ui/views/ItemDelegate.cpp:
```diff
@@ -0,0 +1,260 @@
+	{
+		TreeItem* treeItem = getTreeItemForIndex(index);
+		assert(treeItem != nullptr);
+		if (nullptr != treeItem)
```
assert is disabled when NDEBUG is defined (which it is in the RelWithDebInfo build). So in a RelWithDebInfo build the assert will do nothing. In a debug build (which we use internally), if the assert fires it will print a message and then call abort. That is what we want so we will be forced to investigate why it was null.

seando-adsk commented:
@mattyjams Matt, I've pushed two commits. The first one (https://github.com/Autodesk/maya-usd/pull/206/commits/1115a2e90ed879d3e234685b4959eed0be6de1c9) addresses all your code review comments. The 2nd (https://github.com/Autodesk/maya-usd/pull/206/commits/fac44794450afc107d6af5df205aaa60953f7002) removes all the work-in-progress. I've replied to all your comments. Thanks for having a look at this pull-request. I know it was large, so I really appreciate your time and input.

kxl-adsk commented:
@mattyjams were requested changes made? if so, please approve the PR.

mattyjams commented on lib/fileio/importData.cpp:
```diff
@@ -0,0 +1,146 @@
+{
+	// If we don't have a filename set then we are empty.
+	return fFilename.empty();
+}
```
I could still imagine cases where you might want to import from an open stage as opposed to a file, but I don't think we have any existing usage like that currently. Seems like we have what we need for now.

mattyjams commented on lib/usd/ui/CMakeLists.txt:
```diff
@@ -0,0 +1,122 @@
+
+if (POLICY CMP0074)
+    cmake_policy(SET CMP0074 OLD)
+endif()
```
Still have a policy requirement here?

seando-adsk replied:
Thanks Matt. I remember adding this a while ago to remove some cmake warnings. Since we've now updated cmake min version I'll check these.

seando-adsk replied:
FYI, I looked into these policy statements and I don't think they are needed anymore. There were three places where we were setting 0074 to OLD. I removed all of them and didn't see any cmake warnings. I'll make sure these get removed in a future PR.

mattyjams commented on lib/usd/ui/demo/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+
+if (POLICY CMP0074)
+    cmake_policy(SET CMP0074 OLD)
+endif()
```
Still have a policy requirement here?
